polygon/heimdall/service_observer_options.go

```diff
@@ -0,0 +1,27 @@
+package heimdall
+
+import "math"
+
+func NewObserverOptions(opts ...ObserverOption) ObserverOptions {
+	defaultOptions := ObserverOptions{
+		eventsLimit: math.MaxInt,
+	}
+
+	for _, o := range opts {
+		o(&defaultOptions)
+	}
+
+	return defaultOptions
+}
+
+type ObserverOptions struct {
+	eventsLimit int
+}
+
+type ObserverOption func(opts *ObserverOptions)
+
+func ObserverWithEventsLimit(eventsLimit int) ObserverOption {
```

Comment:
Is it necessary to have all this extra abstraction when all it's being used for is to set the eventsLimit ? Or are you planning to add more complex options in the future?

I would be more in favor of passing the Options directly as a struct, and avoiding this extra complication of defining functions on options, unless you have something concrete in mind.

Reply:
this is a well know Go pattern - https://github.com/uber-go/guide/blob/master/style.md#functional-options
I don't know if I'll be adding more options in the future, but I don't mind the abstraction since it is quite a well know one and it is future proof.

Reply:
Passing in a struct always does not allow me to have a default value `eventsLimit=math.MaxInt`. I just want to have that as default. Check the PR ive linked in the description. There I just do `heimdall.RegisterSpanObserver(callback)` without specifying any options and using the default.

Reply:
Yep, I see . This indeed looks like a more flexible approach.

Reply:
@antonis19 you are definitely right about it maybe being unnecessary for just 1 option right now. I did have a similar doubt but decided not to worry about it now since the abstraction is fairly simple (only 20 lines of code) and also well known and future proof etc.

Reply:
Btw, could we rename this to WithEventsLimit() , since this creates an Option, and not an Observer like the current name suggests.

Reply:
yep, good point, ive renamed it to `WithEventsLimit`